### PR TITLE
[FLINK-9782][e2e] Harden bucketing sink e2e test.

### PR DESCRIPTION
**Pull request by kl0u**
## What is the purpose of the change

Hardens the bucketing sink e2e test. The cause of the instability is described in the JIRA and a travis build where the updated test runs ~45 times can be found here: https://travis-ci.org/kl0u/flink/builds/566401415

## Brief change log

Updated the bucketing sink e2e test.

## Verifying this change

This change is already covered by the existing `BucketingSink` end-to-end test.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-end-to-end-tests/test-scripts/common.sh [hunk 1]
```diff
@@ -553,9 +553,9 @@ function kill_all {
 }
 
 function kill_random_taskmanager {
```
Comment (twalthr) [inline flink-end-to-end-tests/test-scripts/common.sh]: Does this change also other tests? If yes, I would move it to a separate commit.

File: flink-end-to-end-tests/test-scripts/test_streaming_bucketing.sh [hunk 2]
```diff
@@ -20,72 +20,136 @@
...
 
-echo "Starting TM"
-
-# start task manager again
+echo "Restarting 1 TM"
 $FLINK_DIR/bin/taskmanager.sh start
+wait_for_number_of_running_tms 4
 
-echo "Killing 2 TMs"
+sleep 10
```
Comment (twalthr) [inline flink-end-to-end-tests/test-scripts/test_streaming_bucketing.sh]: why is this additional sleep required?

File: flink-end-to-end-tests/flink-bucketing-sink-test/src/main/java/org/apache/flink/streaming/tests/BucketingSinkTestProgram.java [hunk 3]
```diff
@@ -71,7 +71,11 @@ public static void main(String[] args) throws Exception {
 
 		// define bucketing sink to emit the result
 		BucketingSink<Tuple4<Integer, Long, Integer, String>> sink = new BucketingSink<Tuple4<Integer, Long, Integer, String>>(outputPath)
-			.setBucketer(new KeyBucketer());
+				.setBucketer(new KeyBucketer())
+				.setBatchSize(Long.MAX_VALUE)
```
Comment (twalthr) [inline flink-end-to-end-tests/flink-bucketing-sink-test/src/main/java/org/apache/flink/streaming/tests/BucketingSinkTestProgram.java]: What it the reason for these changes? Should the cause of the problem be already solved by having an infinite source now?

File: flink-end-to-end-tests/flink-bucketing-sink-test/src/main/java/org/apache/flink/streaming/tests/BucketingSinkTestProgram.java [hunk 4]
```diff
@@ -143,6 +147,7 @@ public void open(Configuration parameters) {
 		private final int durationMs;
 
 		private long ms = 0;
+		private volatile boolean canceled = false;
```
Comment (twalthr) [inline flink-end-to-end-tests/flink-bucketing-sink-test/src/main/java/org/apache/flink/streaming/tests/BucketingSinkTestProgram.java]: Is the volatile really necessary? I thought every task is single threaded?

Comment (aljoscha) [reply flink-end-to-end-tests/flink-bucketing-sink-test/src/main/java/org/apache/flink/streaming/tests/BucketingSinkTestProgram.java]: There are actually asynchronous calls that are executed from different threads.

File: flink-end-to-end-tests/test-scripts/common.sh [hunk 5]
```diff
@@ -294,7 +294,7 @@ function query_number_of_running_tms {
 
 function wait_for_number_of_running_tms {
   local TM_NUM_TO_WAIT=${1}
-  local TIMEOUT_COUNTER=10
+  local TIMEOUT_COUNTER=30
```
Comment (aljoscha) [inline flink-end-to-end-tests/test-scripts/common.sh]: Maybe revert.

File: flink-end-to-end-tests/test-scripts/test_streaming_bucketing.sh [hunk 6]
```diff
@@ -20,72 +20,136 @@
...
+
+function bucketing_cleanup() {
+  stop_cluster
+  $FLINK_DIR/bin/taskmanager.sh stop-all
+}
+
+# Fix the necessary configuration parameters.
+
+set_conf_ssl
+set_config_key "heartbeat.timeout" "70000"
```
Comment (aljoscha) [inline flink-end-to-end-tests/test-scripts/test_streaming_bucketing.sh]: Double check why this value is as it is now.

File: flink-end-to-end-tests/test-scripts/test_streaming_bucketing.sh (hunk 2 above)
Comment (aljoscha) [reply flink-end-to-end-tests/test-scripts/test_streaming_bucketing.sh]: It's to allow the job to make some progress after killing the TaskManager(s)

File: flink-end-to-end-tests/flink-bucketing-sink-test/src/main/java/org/apache/flink/streaming/tests/BucketingSinkTestProgram.java (hunk 3 above)
Comment (kl0u) [reply flink-end-to-end-tests/flink-bucketing-sink-test/src/main/java/org/apache/flink/streaming/tests/BucketingSinkTestProgram.java]: Actually these are the real instability fix. These make sure that there are no `pending` files between checkpoints which may be invalidated in case of a failure. These files may cause instabilities, given that we _cannot_ safely clean up invalidated `pending` files.

**Comment by kl0u** [comment]
@aljoscha @twalthr I integrated / replied to your comments. Please have a look and let me know what you think.

**Comment by aljoscha** [comment]
Ok, I think this is good to merge now 👌

**Comment by kl0u** [comment]
Thanks a lot for the reviews. I will merge it.
